### ISSUE TITLE
Including header multiple times fails

### ISSUE THREAD

**Issue by dpetican** [EasyG0ing1/BlockNot#12]
I have multiple source code modules each with their own timer.  I don't really understand why but I think the header file violates the [One Definition Rule](https://en.wikipedia.org/wiki/One_Definition_Rule) I moved this code to a separate .cpp file and it now compiles.  I haven't yet done extensive testing.  Thoughts?  BTW I really like your library.  Thanks.

```
#include <BlockNot.h>

void resetAllTimers( const unsigned long newStartTime = millis() ) {
    BlockNot* timer = BlockNot::getFirstTimer();
    while( timer != nullptr ) {
        timer->BlockNot::reset( newStartTime );
        timer = timer->BlockNot::getNextTimer();
    }
}

void resetAllTimers( BlockNot* timer ) { resetAllTimers( timer->BlockNot::getStartTime() ); }

BlockNot* BlockNot::firstTimer   = nullptr;
BlockNot* BlockNot::currentTimer = nullptr;

```
If I understand this issue correctly I think declaring the functions inline in the header file would accomplish the same thing.  BTW is there any problem with declaring each timer instance in each module as static (ie global only to that module)?

**Comment by dpetican** [comment]
Also (BTW I am a C programmer not C++) why do you initialize firstTimer and currentTimer outside of the class definition?

**Comment by EasyG0ing1** [comment]
@dpetican - Hi, thanks for the comments and the issue... the reason why there is a declaration for the library outside of the class definition is that's how we are able to reset all instantiated timers with a single command... 

And if I'm understanding your question properly, there is no problem at all declaring the library as static ... I do it often when I only need a timer to operate within the scope of a single method. 

Perhaps you might be right on the One Definition Rule ... I'd need to run it by someone who understands that better than I do... and maybe it is time to break up the library into a header and a CPP file ... I just kinda wanted to keep it simple as much as possible but it has evolved a great deal since I first wrote it. 

And I'm not really a C or a C++ programmer ... I live in a Java world most of the time and I'm not very fond of C in any flavor but it's a necessary evil in the microcontroller world, unfortunately... of course, Raspberry with their Pico snuck Python into the mix but I don't really see that as a step forward either.

You're the first person who has said that it wouldn't compile (though there may be others who simply didn't bother letting me know) ... I'm curious about your build environment ...

**Comment by dpetican** [comment]
I am using the Visual Studio Code Arduino plugin.  However, I tested this with the standalone Arduino IDE and the compile failed.  On the plus side as I suspected declaring the functions inline solved those linker errors.  But this might lead to code bloat.  I don't get why include guards do not prevent this from happening.

Personally my coding style prefers separate header and source files.  An exception to this is the very well written arduinojson library which is all header using templates.

I see why you initialized firstTimer/currentTimer outside of the class because of this error:

a member with an in-class initializer must be constC/C++(1592)

But don't really understand it.

**Comment by EasyG0ing1** [comment]
@dpetican - Can you provide me with your Arduino IDE settings that generate errors during compiling? Specifically, which board have you selected under the Tools menu? And if it has further settings under the Tools menu, I'll need those as well.

**Comment by dpetican** [comment]
I don't think its my config or board.  Arduino version = 1.8.10 I just tried to replicate the problem by creating a sketch (in Arduino editor) that had 2 modules, each with their own static timers ~~and it compiled fine using a standard Arduino board.~~  Ooops, forgot to use the original library (renamed BlockNot2). Derpy derp!  Okay now it fails with the same errors.  Maybe there is a clue in the "Multiple libraries..." error message?

Error messages:
```
C:\Users\admin\AppData\Local\Temp\arduino_build_691564\sketch\blink2.cpp.o (symbol from plugin): In function `BlockNot::firstTimer':
(.text+0x0): multiple definition of `BlockNot::firstTimer'
C:\Users\admin\AppData\Local\Temp\arduino_build_691564\sketch\blink1.cpp.o (symbol from plugin):(.text+0x0): first defined here
c:/program files (x86)/arduino/hardware/tools/avr/bin/../lib/gcc/avr/7.3.0/../../../../avr/bin/ld.exe: Disabling relaxation: it will not work with multiple definitions
C:\Users\admin\AppData\Local\Temp\arduino_build_691564\sketch\blink2.cpp.o (symbol from plugin): In function `BlockNot::firstTimer':
(.text+0x0): multiple definition of `BlockNot::currentTimer'
C:\Users\admin\AppData\Local\Temp\arduino_build_691564\sketch\blink1.cpp.o (symbol from plugin):(.text+0x0): first defined here
C:\Users\admin\AppData\Local\Temp\arduino_build_691564\sketch\blink2.cpp.o (symbol from plugin): In function `BlockNot::firstTimer':
(.text+0x0): multiple definition of `resetAllTimers(unsigned long)'
C:\Users\admin\AppData\Local\Temp\arduino_build_691564\sketch\blink1.cpp.o (symbol from plugin):(.text+0x0): first defined here
C:\Users\admin\AppData\Local\Temp\arduino_build_691564\sketch\blink2.cpp.o (symbol from plugin): In function `BlockNot::firstTimer':
(.text+0x0): multiple definition of `resetAllTimers(BlockNot*)'
C:\Users\admin\AppData\Local\Temp\arduino_build_691564\sketch\blink1.cpp.o (symbol from plugin):(.text+0x0): first defined here
collect2.exe: error: ld returned 1 exit status
Multiple libraries were found for "BlockNot2.h"
Used: C:\Users\admin\Documents\Arduino\libraries\BlockNot2
Using library BlockNot2 at version 1.7.2 in folder: C:\Users\admin\Documents\Arduino\libraries\BlockNot2 
exit status 1
Error compiling for board Arduino/Genuino Mega or Mega 2560.
```

BlockNotBlink.ino:
```
#include "blink1.h"
#include "blink2.h"

void setup() {
  pinMode(9, OUTPUT);
  pinMode(10, OUTPUT);
}

void loop() {
  blink1();
  blink2();
}

```
blink1.h:
```
#ifndef BLINK1_H_
#define BLINK1_H_

void blink1(void);

#endif
```
blink1.cpp:
```
#include <BlockNot2.h>
#include "blink1.h"

static BlockNot blinkTimer(100);
static boolean state = false;

void blink1(void) {
  if (blinkTimer.TRIGGERED) {
      state = !state;
      digitalWrite(9, (state ? HIGH : LOW));
  }
}
```
blink2.h:
```
#ifndef BLINK2_H_
#define BLINK2_H_

void blink2(void);

#endif
```
blink2.cpp:
```
#include <BlockNot2.h>
#include "blink2.h"

static BlockNot blinkTimer(100);
static boolean state = false;

void blink2(void) {
  if (blinkTimer.TRIGGERED) {
      state = !state;
      digitalWrite(10, (state ? HIGH : LOW));
  }
}
```

**Comment by EasyG0ing1** [comment]
@dpetican   The latest version of the Arduino IDE is 1.8.15 - install that version and try your compile again. Also, I need to know what you're configured to compile for ... you can find that in your TOOLS menu [as seen in this pic](https://imgur.com/G781PqZ) ... 

And a full copy of the sketch that is failing would be helpful too ... I need to re-create the problem so that I can fix it.

Thanks,

Mike

**Comment by dpetican** [comment]
The full sketch that is failing is what I included above.  The _real_ program as it were is way too big and complex to be useful in this discussion.  Note that I included BlockNot2.h This is for me the unmodified version of the library.  Change to BlockNot.h for your test.  I updated to 1.8.15 and get essentially the same errors as before.  On the other hand my boards menu (with many cores) is now easier to navigate and the programmer sub-menu changes with the core selected.  Thanks for that nudge.

Hmm interesting.  Several warning? messages from the previous Arduino 1.8.10 have gone away.  I'm trying to migrate to using the Arduino CLI so I have more control over some of these linker and warning things, but I'm having problems getting it to work with VS Code.

```
c:/program files (x86)/arduino/hardware/tools/avr/bin/../lib/gcc/avr/7.3.0/../../../../avr/bin/ld.exe: Disabling relaxation: it will not work with multiple definitions
Multiple libraries were found for "BlockNot2.h"
Used: C:\Users\admin\Documents\Arduino\libraries\BlockNot2
```

Full error messages from Arduino 1.8.15:
```
C:\Users\admin\AppData\Local\Temp\arduino_build_590445\sketch\blink2.cpp.o (symbol from plugin): In function `BlockNot::firstTimer':
(.text+0x0): multiple definition of `BlockNot::firstTimer'
C:\Users\admin\AppData\Local\Temp\arduino_build_590445\sketch\blink1.cpp.o (symbol from plugin):(.text+0x0): first defined here
C:\Users\admin\AppData\Local\Temp\arduino_build_590445\sketch\blink2.cpp.o (symbol from plugin): In function `BlockNot::firstTimer':
(.text+0x0): multiple definition of `BlockNot::currentTimer'
C:\Users\admin\AppData\Local\Temp\arduino_build_590445\sketch\blink1.cpp.o (symbol from plugin):(.text+0x0): first defined here
C:\Users\admin\AppData\Local\Temp\arduino_build_590445\sketch\blink2.cpp.o (symbol from plugin): In function `BlockNot::firstTimer':
(.text+0x0): multiple definition of `resetAllTimers(unsigned long)'
C:\Users\admin\AppData\Local\Temp\arduino_build_590445\sketch\blink1.cpp.o (symbol from plugin):(.text+0x0): first defined here
C:\Users\admin\AppData\Local\Temp\arduino_build_590445\sketch\blink2.cpp.o (symbol from plugin): In function `BlockNot::firstTimer':
(.text+0x0): multiple definition of `resetAllTimers(BlockNot*)'
C:\Users\admin\AppData\Local\Temp\arduino_build_590445\sketch\blink1.cpp.o (symbol from plugin):(.text+0x0): first defined here
collect2.exe: error: ld returned 1 exit status
Using library BlockNot2 at version 1.7.2 in folder: C:\Users\admin\Documents\Arduino\libraries\BlockNot2 
exit status 1
Error compiling for board Arduino Mega or Mega 2560.

```

![image](https://user-images.githubusercontent.com/56268293/129789977-e0e5fc4a-cfe0-4c36-8599-73ce0b216bc6.png)

![image](https://user-images.githubusercontent.com/56268293/129791595-f43580e3-0ff6-47f3-b2a3-755e4775f8ca.png)

**Comment by EasyG0ing1** [comment]
@dpetican  - off topic, but have you looked at the CLion IDE? I use it with Platform IO and it is quite ... magical as IDEs go...

**Comment by EasyG0ing1** [comment]
@dpetican - OK, I see the same problem that you're seeing, and you'll have to forgive me as I simply don't know enough about the C language in general - there is much about it that I don't understand nor have I really tried to understand. I've learned what I needed to learn to do the things I've needed to do. 

So, for example, I have of course seen libraries where they are written using .cpp files and .h files, but I never really understood the full relationship between them, nor have I really thought about the reason why C (as a language class) uses two different files. It is obvious, however, given the problem here, that I need more knowledge.

I'm going to spend some time today educating myself to see if I can get a handle on this error and then fix the library.

Edit: I found and read [this file](http://websites.umich.edu/~eecs381/handouts/CppHeaderFileGuidelines.pdf) ... which seems quite relevant in this situation. 
This also appears to be a [very useful site](https://www.learncpp.com/):

**Comment by EasyG0ing1** [comment]
@dpetican - OK looks like I got the library working and compiling just fine using your two header example ... I just need to test the functionality of that resetAll method and if it works, then I'll clean everything up and push the updated code to the repository. I did end up breaking the library up into a header and a code file.

**Comment by EasyG0ing1** [comment]
@dpetican - OK, I have the library working properly now and your example compiles and functions properly as does the global reset of all timers method... a famous line from an old famous movie, "This house ... is clean..."  :-)

I'm going to give the documentation a once over before I publish and I gave you a credit in the version notes at the bottom of the readme for spotting the problem.

**Comment by EasyG0ing1** [comment]
Issue fixed in 1.8.0 update

**Comment by dpetican** [comment]
Sweet. Thanks.
